engine/lib/filestore.php

```diff
@@ -598,6 +578,209 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
+	$response->prepare($request);
+
+	$guid = array_shift($segments);
+	$size = array_shift($segments);
+
+	$file = get_entity($guid);
+	if (!$file instanceof ElggFile) {
+		if (elgg_entity_exists($guid)) {
+			$response->setStatusCode(403)->setContent('You are not allowed to access this file')->send();
+		} else {
+			$response->setStatusCode(404)->setContent('File does not exist')->send();
+		}
+		exit;
+	}
+
+	$thumbnail = elgg_get_thumbnail($file, $size);
+	if (!$thumbnail->exists()) {
+		$response->setStatusCode(404)->setContent('Thumbnail doest not exist')->send();
+		exit;
+	}
+
+	$if_none_match = $request->headers->get('if_none_match');
+	if (!empty($if_none_match)) {
+		// strip mod_deflate suffixes
+		$request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+	}
+
+	$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+	$last_updated = filemtime($filenameonfilestore);
+	$etag = '"' . $last_updated . '"';
+	$response->setPublic()->setEtag($etag);
+	if ($response->isNotModified($request)) {
+		$response->send();
+		exit();
+	}
+
+	header_remove('Cache-Control');
+	header_remove('Pragma');
+	header_remove('Expires');
+	
+	$headers = [
+		'Content-Type' => (new Elgg\Filesystem\MimeTypeDetector())->getType($filenameonfilestore),
+	];
+	$response = new Symfony\Component\HttpFoundation\BinaryFileResponse($filenameonfilestore, 200, $headers, false, 'inline');
+	$response->prepare($request);
+
+	$expires_dt = (new DateTime())->setTimestamp(strtotime('+1 day'));
+	$response->setExpires($expires_dt);
+
+	$response->setEtag($etag);
+	
+	$response->send();
+	exit;
+}
+
+/**
+ * Returns thumbnail sizes configuration for a file entity
+ *
+ * @param \ElggFile $file File entity
+ * @return array
+ */
+function elgg_get_thumbnail_sizes(\ElggFile $file) {
+
+	$sizes = [
+		'small' => [
+			'w' => 60,
+			'h' => 60,
+			'square' => true,
+			'metadata_name' => 'thumbnail',
+			'filename_prefix' => 'thumb',
+		],
+		'medium' => [
+			'w' => 153,
+			'h' => 153,
+			'square' => true,
+			'metadata_name' => 'smallthumb',
+			'filename_prefix' => 'smallthumb',
+		],
+		'large' => [
+			'w' => 600,
+			'h' => 600,
+			'square' => false,
+			'metadata_name' => 'largethumb',
+			'filename_prefix' => 'largethumb',
+		],
+	];
+
+	// @todo: add a hook to filter these values
+	return $sizes;
+}
+
+/**
+ * Create thumbnails for an image file
+ *
+ * @param \ElggFile $file File entity
+ * @return bool
+ */
+function elgg_create_thumbnails(\ElggFile $file) {
+
+	$prefix = 'file/';
+	$filestorename = elgg_substr($file->getFilename(), elgg_strlen($prefix));
+
+	$sizes = elgg_get_thumbnail_sizes($file);
+
+	foreach ($sizes as $size => $opts) {
+		$metadata_name = $opts['metadata_name'];
+		$filename_prefix = $opts['filename_prefix'];
+		$width = $opts['w'];
+		$height = $opts['h'];
+		$square = $opts['square'];
+		if (!isset($file->$metadata_name)) {
+			$file->$metadata_name = $prefix . $filename_prefix . $filestorename;
```

Comment:
I'm OK duplicating flawed logic for the meantime, but we should make a ticket and a `@todo` in comments linking to the ticket so we can fix later.